lib/facility.rb

```diff
@@ -1,14 +1,142 @@
+# custom errors site link = https://www.honeybadger.io/blog/ruby-custom-exceptions/
+
+class WrittenTestAlreadyTakenError < StandardError
+
+  attr_reader :message
+  def initialize(msg="User has already taken the written test")
+    @message = msg
+    super(@message)
+  end
+end
+
+class RoadTestAlreadyTakenError < StandardError
+  attr_reader :message
+  def initialize(msg="User has already taken the road test")
+    @message = msg
+    super(@message)
+  end
+end
+
+class VehicleAlreadyRegisted < StandardError
+  attr_reader :message
+  def initialize(msg="This vehicle is already registered")
+    @message = msg
+    super(@message)
+  end
+end
+
 class Facility
-  attr_reader :name, :address, :phone, :services
+  attr_accessor :name, :address, :phone, :services, :registered_vehicles, :collected_fees, :holidays, :daily_hours
 
-  def initialize(name, address, phone)
-    @name = name
-    @address = address
-    @phone = phone
+  def initialize(info)
+    info.default = "not availabe"
+    @name = info[:name]
+    @address = info[:address]
+    @phone = info[:phone]
+    @holidays = info[:holidays] # iteration 4
+    @daily_hours = info[:daily_hours] #iteration 4
     @services = []
+    @registered_vehicles = []
+    @collected_fees = 0
   end
 
-  def add_services(service)
+  def add_service(service)
     @services << service
   end
+
+
+
+  def register_vehicle(vehicle)
+
+    if @services.include?('Vehicle Registration')
+      raise TypeError, "Vehicle must be a vehicle object" unless vehicle.is_a?(Vehicle)
+      
+      existing_vehicle_check = registered_vehicles.find { |existing_vehicle| existing_vehicle.vin == vehicle.vin}
+      if existing_vehicle_check
+        raise VehicleAlreadyRegisted.new()
+      end
+
+      if vehicle.electric_vehicle?
+        vehicle.set_platetype(:ev)
+        @collected_fees += 200
+
+      elsif vehicle.antique?
+        vehicle.set_platetype(:antique)
+        @collected_fees += 25
+
+      else
+        vehicle.set_platetype(:regular)
+        @collected_fees += 100
+
+      end
+
+      vehicle.set_registration_date
+
+      return @registered_vehicles << vehicle
+
+      # return "#{vehicle.model} has been registered"
+
+    else
+      return @registered_vehicles
+    end
+
+  end
+
+  def administer_written_test(registrant)
+    if @services.include?('Written Test')
+      raise TypeError, "registrant must be a registrant object" unless registrant.is_a?(Registrant)
+      raise WrittenTestAlreadyTakenError.new() unless registrant.license_data[:written] == false
```

Comment:
In Ruby, we do not use open parenthesis to invoke a method, so just `.new` not `.new()`